File: cause-for-concern/prototype/cfc/kernel.py
"""
cfc.kernel -- the numerics behind the language.

Implements, with no dependency beyond the standard library:

  * circuit construction   mu = mu0 + RT ln c,  G = k c / RT,  J = G dmu
  * gauge centring         (Prop. "potential offsets are a gauge freedom")
  * cycle bases            fundamental (spanning tree) and a shortest-
                           cycle greedy basis
  * cycle sums             the quantity that is identically zero
  * numerical floor        eps_num = gamma_{2L} * 2 L Lambda
  * data floor             eps_data = z * sqrt(sum sigma_i^2)
  * the three-valued verdict and the witness set

The two floors are kept apart everywhere, because collapsing them is
exactly the error the companion paper is about.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Sequence, Set, Tuple

# Unit roundoff for IEEE-754 binary64.
MACH_U = 2.0 ** -53

# Gas constant times body temperature, kJ/mol at T = 310 K.
RT_DEFAULT = 2.577

# Normal quantile for a two-sided interval at alpha = 0.01.
Z_ALPHA = 2.5758293035489004


class KernelError(Exception):
    pass


# =====================================================================
# Circuit
# =====================================================================

@dataclass
class Species:
    name: str
    mu0: float
    concentration: float
    sigma: Optional[float] = None       # uncertainty on mu, kJ/mol


@dataclass
class Reaction:
    name: str
    src: str
    dst: str
    k: float


@dataclass
class Circuit:
    species: Dict[str, Species] = field(default_factory=dict)
    reactions: List[Reaction] = field(default_factory=list)
    RT: float = RT_DEFAULT

    # populated by solve()
    mu: Dict[str, float] = field(default_factory=dict)
    G: Dict[str, float] = field(default_factory=dict)
    J: Dict[str, float] = field(default_factory=dict)
    solved: bool = False
    centred: bool = False
    gauge_offset: float = 0.0

    # Per-edge offsets applied to the potential difference of a
    # reaction, i.e. d_e = (mu_dst - mu_src) + offset_e.
    #
    # An offset makes the edge data fail to be a gradient, which IS
    # what thermodynamic inconsistency means. Perturbing a NODE
    # potential cannot do this: any change to mu leaves the data a
    # gradient of the changed mu, so every cycle sum stays exactly
    # zero. The defect has to live on an edge.
    edge_offset: Dict[str, float] = field(default_factory=dict)

    # ---- construction ------------------------------------------------

    def add_species(self, s: Species) -> None:
        if s.concentration <= 0:
            raise KernelError(
                f"species {s.name}: concentration must be > 0")
        self.species[s.name] = s

    def add_reaction(self, r: Reaction) -> None:
        for end in (r.src, r.dst):
            if end not in self.species:
                raise KernelError(
                    f"reaction {r.name}: unknown species {end!r}")
        if r.k <= 0:
            raise KernelError(f"reaction {r.name}: k must be > 0")
        self.reactions.append(r)

    # ---- solving -----------------------------------------------------

    def solve(self) -> "Circuit":
        """Compute mu, G, J.  Pure evaluation; no iteration required."""
        self.mu = {
            name: s.mu0 + self.RT * math.log(s.concentration)
            for name, s in self.species.items()
        }
        self.G, self.J = {}, {}
        for r in self.reactions:
            g = r.k * self.species[r.src].concentration / self.RT
            self.G[r.name] = g
            self.J[r.name] = g * (self.mu[r.src] - self.mu[r.dst])
        self.solved = True
        return self

    def copy(self) -> "Circuit":
        c = Circuit(
            species={k: Species(v.name, v.mu0, v.concentration, v.sigma)
                     for k, v in self.species.items()},
            reactions=[Reaction(r.name, r.src, r.dst, r.k)
                       for r in self.reactions],
            RT=self.RT,
            edge_offset=dict(self.edge_offset),
        )
        return c

    def delta_mu(self, r: Reaction) -> float:
        """Edge data d_e as supplied: gradient part plus any offset."""
        return (self.mu[r.dst] - self.mu[r.src]) + self.edge_offset.get(
            r.name, 0.0)

    # ---- gauge -------------------------------------------------------

    def centre_potentials(self) -> "Circuit":
        """Shift all potentials so max|mu| is minimised.

        Changes no cycle sum mathematically; reduces the numerical floor
        because rounding error scales with operand magnitude.
        """
        if not self.solved:
            self.solve()
        out = self.copy()
        out.mu = dict(self.mu)
        out.G, out.J = dict(self.G), dict(self.J)
        out.solved = True
        if out.mu:
            lo, hi = min(out.mu.values()), max(out.mu.values())
            offset = -0.5 * (lo + hi)
            out.mu = {k: v + offset for k, v in out.mu.items()}
            out.gauge_offset = offset
        out.centred = True
        return out

    # ---- topology ----------------------------------------------------

    def node_balance(self) -> Dict[str, float]:
        """Signed flux sum at each species (Kirchhoff current law)."""
        bal = {name: 0.0 for name in self.species}
        for r in self.reactions:
            bal[r.src] -= self.J[r.name]
            bal[r.dst] += self.J[r.name]
        return bal

    def cyclomatic_number(self) -> int:
        n = len(self.species)
        m = len(self.reactions)
        comps = _count_components(self.species.keys(), self.reactions)
        return m - n + comps


def _count_components(nodes, reactions) -> int:
    parent = {n: n for n in nodes}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for r in reactions:
        a, b = find(r.src), find(r.dst)
        if a != b:
            parent[a] = b
    return len({find(n) for n in nodes})


# =====================================================================
# Cycles
# =====================================================================

@dataclass
class Cycle:
    """A cycle as a signed list of (reaction_name, orientation)."""
    name: str
    edges: List[Tuple[str, int]]        # (+1 forward, -1 reversed)
    nodes: List[str]

    @property
    def length(self) -> int:
        return len(self.edges)

    def edge_names(self) -> Set[str]:
        return {e for e, _ in self.edges}


def fundamental_cycle_basis(c: Circuit) -> List[Cycle]:
    """Cycles from the fundamental system of a spanning forest."""
    adj: Dict[str, List[Tuple[str, str]]] = {n: [] for n in c.species}
    for r in c.reactions:
        adj[r.src].append((r.dst, r.name))
        adj[r.dst].append((r.src, r.name))

    parent: Dict[str, Optional[Tuple[str, str]]] = {}
    seen: Set[str] = set()
    tree_edges: Set[str] = set()

    for root in c.species:
        if root in seen:
            continue
        seen.add(root)
        parent[root] = None
        stack = [root]
        while stack:
            u = stack.pop()
            for v, ename in adj[u]:
                if v not in seen:
                    seen.add(v)
                    parent[v] = (u, ename)
                    tree_edges.add(ename)
                    stack.append(v)

    by_name = {r.name: r for r in c.reactions}
    cycles: List[Cycle] = []
    for r in c.reactions:
        if r.name in tree_edges:
            continue
        path = _tree_path(r.dst, r.src, parent, by_name)
        if path is None:
            continue
        edges = [(r.name, +1)] + path
        nodes = _nodes_of(edges, by_name, start=r.src)
        cycles.append(Cycle(name=f"cyc_{r.name}", edges=edges, nodes=nodes))
    return cycles


def _tree_path(a: str, b: str, parent, by_name) -> Optional[List[Tuple[str, int]]]:
    """Signed edge path from a to b through the spanning forest."""
    anc_a, node = [], a
    seen_a = {a: 0}
    while parent.get(node):
        u, ename = parent[node]
        anc_a.append((ename, node, u))
        node = u
        seen_a[node] = len(anc_a)

    path_b, node = [], b
    while node not in seen_a:
        p = parent.get(node)
        if p is None:
            return None
        u, ename = p
        path_b.append((ename, node, u))
        node = u
    meet = node

    out: List[Tuple[str, int]] = []
    for ename, child, par in anc_a[:seen_a[meet]]:
        r = by_name[ename]
        out.append((ename, +1 if r.src == child else -1))
    for ename, child, par in reversed(path_b):
        r = by_name[ename]
        out.append((ename, +1 if r.src == par else -1))
    return out


def _nodes_of(edges, by_name, start: str) -> List[str]:
    nodes, cur = [start], start
    for ename, sign in edges:
        r = by_name[ename]
        nxt = r.dst if (r.src == cur) else r.src
        nodes.append(nxt)
        cur = nxt
    return nodes


def minimum_cycle_basis(c: Circuit) -> List[Cycle]:
    """Greedy shortest-cycle basis.

    For each non-tree edge, take the shortest cycle through it (BFS in
    the graph minus that edge). Shorter cycles intersect in fewer edges,
    which sharpens the witness set; that is the reason to prefer this
    over the fundamental basis.
    """
    by_name = {r.name: r for r in c.reactions}
    nu = c.cyclomatic_number()
    if nu <= 0:
        return []

    candidates: List[Cycle] = []
    for r in c.reactions:
        path = _shortest_path_excluding(c, r.dst, r.src, exclude=r.name)
        if path is None:
            continue
        edges = [(r.name, +1)] + path
        nodes = _nodes_of(edges, by_name, start=r.src)
        candidates.append(Cycle(name=f"cyc_{r.name}", edges=edges, nodes=nodes))

    candidates.sort(key=lambda cy: (cy.length, cy.name))

    # Greedily keep cycles independent over GF(2).
    ordered = [r.name for r in c.reactions]
    idx = {e: i for i, e in enumerate(ordered)}
    basis_vecs: List[int] = []
    chosen: List[Cycle] = []
    for cy in candidates:
        vec = 0
        for ename, _ in cy.edges:
            vec ^= 1 << idx[ename]
        reduced = vec
        for bv in basis_vecs:
            reduced = min(reduced, reduced ^ bv)
        if reduced:
            basis_vecs.append(reduced)
            basis_vecs.sort(reverse=True)
            chosen.append(cy)
        if len(chosen) == nu:
            break
    return chosen


def _shortest_path_excluding(c: Circuit, a: str, b: str,
                             exclude: str) -> Optional[List[Tuple[str, int]]]:
    from collections import deque
    adj: Dict[str, List[Tuple[str, str]]] = {n: [] for n in c.species}
    for r in c.reactions:
        if r.name == exclude:
            continue
        adj[r.src].append((r.dst, r.name))
        adj[r.dst].append((r.src, r.name))

    prev: Dict[str, Optional[Tuple[str, str]]] = {a: None}
    q = deque([a])
    while q:
        u = q.popleft()
        if u == b:
            break
        for v, ename in adj[u]:
            if v not in prev:
                prev[v] = (u, ename)
                q.append(v)
    if b not in prev:
        return None

    by_name = {r.name: r for r in c.reactions}
    out: List[Tuple[str, int]] = []
    node = b
    while prev[node] is not None:
        u, ename = prev[node]
        r = by_name[ename]
        out.append((ename, +1 if r.src == u else -1))
        node = u
    out.reverse()
    return out


# =====================================================================
# Cycle sums and the two floors
# =====================================================================

def cycle_sum(c: Circuit, cy: Cycle) -> float:
    """Sum of edge data around `cy`.

    If the edge data are a gradient -- i.e. every edge offset is zero --
    this telescopes to exactly zero for any potentials, so what is
    returned is pure accumulated rounding error. A nonzero edge offset
    breaks the gradient property, and its net signed traversal count
    around the cycle is what survives.
    """
    by_name = {r.name: r for r in c.reactions}
    total = 0.0
    for ename, sign in cy.edges:
        total += sign * c.delta_mu(by_name[ename])
    return total


def potential_range(c: Circuit, cy: Cycle) -> float:
    """Lambda: the largest absolute potential on the cycle."""
    return max(abs(c.mu[n]) for n in set(cy.nodes)) if cy.nodes else 0.0


def gamma(k: int, u: float = MACH_U) -> float:
    denom = 1.0 - k * u
    if denom <= 0:
        raise KernelError("gamma: k*u >= 1, error bound unusable")
    return (k * u) / denom


def numerical_floor(c: Circuit, cy: Cycle) -> float:
    """eps_num = gamma_{2L} * 2 L Lambda."""
    L = cy.length
    lam = potential_range(c, cy)
    return gamma(2 * L) * 2.0 * L * lam


def data_floor(c: Circuit, cy: Cycle,
               z: float = Z_ALPHA) -> Optional[float]:
    """eps_data = z * sqrt(sum_i sigma_i^2), or None if any sigma absent.

    Returning None rather than substituting a default is deliberate: a
    fabricated uncertainty converts an unanswerable question into a
    confident answer.
    """
    total = 0.0
    for name in set(cy.nodes):
        s = c.species[name].sigma
        if s is None:
            return None
        total += s * s
    return z * math.sqrt(total)


# =====================================================================
# Verdict
# =====================================================================

CONSISTENT = "CONSISTENT"
UNDECIDABLE = "UNDECIDABLE"
INCONSISTENT = "INCONSISTENT"


@dataclass
class Tolerance:
    numerical: float
    data: Optional[float]
    star: float
    loop: str

    def to_json(self) -> dict:
        return {
            "loop": self.loop,
            "numerical": self.numerical,
            "data": self.data,
            "star": self.star,
            "data_available": self.data is not None,
        }


@dataclass
class Holonomy:
    value: float
    loop: str
    length: int
    potential_range: float

    def to_json(self) -> dict:
        return {
            "loop": self.loop,
            "value": self.value,
            "abs_value": abs(self.value),
            "length": self.length,
            "potential_range": self.potential_range,
        }


@dataclass
class Verdict:
    """A verdict ALWAYS carries the tolerance that produced it."""
    label: str
    holonomy: Holonomy
    tolerance: Tolerance

    def to_json(self) -> dict:
        return {
            "verdict": self.label,
            "holonomy": self.holonomy.to_json(),
            "tolerance": self.tolerance.to_json(),
        }

    def __eq__(self, other) -> bool:
        if isinstance(other, str):
            return self.label == other
        if isinstance(other, Verdict):
            return self.label == other.label
        return NotImplemented

    def __hash__(self):
        return hash(self.label)


def compute_tolerance(c: Circuit, cy: Cycle,
                      z: float = Z_ALPHA) -> Tolerance:
    en = numerical_floor(c, cy)
    ed = data_floor(c, cy, z=z)
    star = en if ed is None else max(en, ed)
    return Tolerance(numerical=en, data=ed, star=star, loop=cy.name)


def compute_holonomy(c: Circuit, cy: Cycle) -> Holonomy:
    return Holonomy(value=cycle_sum(c, cy), loop=cy.name,
                    length=cy.length, potential_range=potential_range(c, cy))


def admit(h: Holonomy, t: Tolerance) -> Verdict:
    """The three-valued classification.

    Below the numerical floor          -> CONSISTENT
    Between numerical and data floor   -> UNDECIDABLE
    Above both                         -> INCONSISTENT
    """
    a = abs(h.value)
    if a <= t.numerical:
        label = CONSISTENT
    elif t.data is not None and a <= t.data:
        label = UNDECIDABLE
    elif t.data is None and a > t.numerical:
        # No uncertainty data: cannot distinguish real defect from
        # unquantified data error.  Honest answer is UNDECIDABLE.
        label = UNDECIDABLE
    else:
        label = INCONSISTENT
    return Verdict(label=label, holonomy=h, tolerance=t)


def witness_set(flagged: Sequence[Cycle]) -> Set[str]:
    """Edges lying on EVERY flagged cycle.

    The basis-independent content of a positive result: the defect is in
    here. Naming a single loop is not basis-independent.
    """
    if not flagged:
        return set()
    out = set(flagged[0].edge_names())
    for cy in flagged[1:]:
        out &= cy.edge_names()
    return out

File: cause-for-concern/webtool/src/cfc/kernel.js
/**
 * cfc/kernel.js -- the numerics, ported from the Python reference
 * implementation in ../../prototype/cfc/kernel.py.
 *
 * This is a real implementation, not a mock. Circuits are solved,
 * cycle bases are computed, and both tolerance floors are derived from
 * the same formulae the paper states:
 *
 *   mu_i    = mu0_i + RT ln c_i
 *   G_ij    = k_ij c_i / RT
 *   J_ij    = G_ij (mu_i - mu_j)
 *   eps_num = gamma_{2L} * 2 L Lambda        (Thm 3.2)
 *   eps_dat = z * sqrt(sum_i sigma_i^2)      (Thm 4.2)
 *
 * The cycle sum is identically zero for consistent edge data, so a
 * nonzero value is either rounding error or a genuine defect. Telling
 * those apart is the whole point.
 */

export const MACH_U = Math.pow(2, -53);      // IEEE-754 binary64
export const RT_DEFAULT = 2.577;             // kJ/mol at T = 310 K
export const Z_ALPHA = 2.5758293035489004;   // two-sided, alpha = 0.01

export const CONSISTENT = "CONSISTENT";
export const UNDECIDABLE = "UNDECIDABLE";
export const INCONSISTENT = "INCONSISTENT";

export class KernelError extends Error {}

// ---------------------------------------------------------------------
// Circuit
// ---------------------------------------------------------------------

export class Circuit {
  constructor(RT = RT_DEFAULT) {
    this.species = new Map();     // name -> {name, mu0, concentration, sigma}
    this.reactions = [];          // {name, src, dst, k}
    this.RT = RT;
    this.mu = new Map();
    this.G = new Map();
    this.J = new Map();
    this.edgeOffset = new Map();  // reaction name -> offset on d_e
    this.solved = false;
    this.centred = false;
    this.gaugeOffset = 0;
  }

  addSpecies(s) {
    if (!(s.concentration > 0)) {
      throw new KernelError(`species ${s.name}: concentration must be > 0`);
    }
    this.species.set(s.name, { sigma: null, ...s });
  }

  addReaction(r) {
    for (const end of [r.src, r.dst]) {
      if (!this.species.has(end)) {
        throw new KernelError(`reaction ${r.name}: unknown species '${end}'`);
      }
    }
    if (!(r.k > 0)) throw new KernelError(`reaction ${r.name}: k must be > 0`);
    const existing = this.reactions.find((x) => x.name === r.name);
    if (existing) Object.assign(existing, r);
    else this.reactions.push({ ...r });
  }

  solve() {
    this.mu = new Map();
    for (const [name, s] of this.species) {
      this.mu.set(name, s.mu0 + this.RT * Math.log(s.concentration));
    }
    this.G = new Map();
    this.J = new Map();
    for (const r of this.reactions) {
      const g = (r.k * this.species.get(r.src).concentration) / this.RT;
      this.G.set(r.name, g);
      this.J.set(r.name, g * (this.mu.get(r.src) - this.mu.get(r.dst)));
    }
    this.solved = true;
    return this;
  }

  copy() {
    const c = new Circuit(this.RT);
    for (const [k, v] of this.species) c.species.set(k, { ...v });
    c.reactions = this.reactions.map((r) => ({ ...r }));
    c.edgeOffset = new Map(this.edgeOffset);
    c.mu = new Map(this.mu);
    c.G = new Map(this.G);
    c.J = new Map(this.J);
    c.solved = this.solved;
    c.centred = this.centred;
    c.gaugeOffset = this.gaugeOffset;
    return c;
  }

  /** Edge data as supplied: gradient part plus any injected offset. */
  deltaMu(r) {
    return (
      this.mu.get(r.dst) - this.mu.get(r.src) + (this.edgeOffset.get(r.name) || 0)
    );
  }

  /**
   * Shift all potentials so max|mu| is minimised. Changes no cycle sum
   * mathematically; lowers the numerical floor, because rounding error
   * scales with operand magnitude rather than with their difference.
   */
  centrePotentials() {
    if (!this.solved) this.solve();
    const out = this.copy();
    const vals = [...out.mu.values()];
    if (vals.length) {
      const lo = Math.min(...vals);
      const hi = Math.max(...vals);
      const off = -0.5 * (lo + hi);
      for (const [k, v] of out.mu) out.mu.set(k, v + off);
      out.gaugeOffset = off;
    }
    out.centred = true;
    return out;
  }

  nodeBalance() {
    const bal = new Map();
    for (const n of this.species.keys()) bal.set(n, 0);
    for (const r of this.reactions) {
      bal.set(r.src, bal.get(r.src) - this.J.get(r.name));
      bal.set(r.dst, bal.get(r.dst) + this.J.get(r.name));
    }
    return bal;
  }

  cyclomaticNumber() {
    const n = this.species.size;
    const m = this.reactions.length;
    return m - n + countComponents(this.species.keys(), this.reactions);
  }
}

function countComponents(nodes, reactions) {
  const parent = new Map();
  for (const n of nodes) parent.set(n, n);
  const find = (x) => {
    while (parent.get(x) !== x) {
      parent.set(x, parent.get(parent.get(x)));
      x = parent.get(x);
    }
    return x;
  };
  for (const r of reactions) {
    const a = find(r.src);
    const b = find(r.dst);
    if (a !== b) parent.set(a, b);
  }
  const roots = new Set();
  for (const n of parent.keys()) roots.add(find(n));
  return roots.size;
}

// ---------------------------------------------------------------------
// Cycles
// ---------------------------------------------------------------------

export class Cycle {
  constructor(name, edges, nodes) {
    this.name = name;
    this.edges = edges;        // [[reactionName, +1|-1], ...]
    this.nodes = nodes;
  }
  get length() {
    return this.edges.length;
  }
  edgeNames() {
    return new Set(this.edges.map(([e]) => e));
  }
}

function adjacency(c, exclude = null) {
  const adj = new Map();
  for (const n of c.species.keys()) adj.set(n, []);
  for (const r of c.reactions) {
    if (r.name === exclude) continue;
    adj.get(r.src).push([r.dst, r.name]);
    adj.get(r.dst).push([r.src, r.name]);
  }
  return adj;
}

function nodesOf(edges, byName, start) {
  const nodes = [start];
  let cur = start;
  for (const [ename] of edges) {
    const r = byName.get(ename);
    const nxt = r.src === cur ? r.dst : r.src;
    nodes.push(nxt);
    cur = nxt;
  }
  return nodes;
}

function shortestPathExcluding(c, a, b, exclude) {
  const adj = adjacency(c, exclude);
  const prev = new Map([[a, null]]);
  const q = [a];
  let qi = 0;
  while (qi < q.length) {
    const u = q[qi++];
    if (u === b) break;
    for (const [v, ename] of adj.get(u) || []) {
      if (!prev.has(v)) {
        prev.set(v, [u, ename]);
        q.push(v);
      }
    }
  }
  if (!prev.has(b)) return null;
  const byName = new Map(c.reactions.map((r) => [r.name, r]));
  const out = [];
  let node = b;
  while (prev.get(node) !== null) {
    const [u, ename] = prev.get(node);
    const r = byName.get(ename);
    out.push([ename, r.src === u ? +1 : -1]);
    node = u;
  }
  out.reverse();
  return out;
}

/**
 * Greedy shortest-cycle basis. Shorter cycles intersect in fewer edges,
 * which sharpens the witness set -- the reason to prefer this over the
 * fundamental basis (paper, Remark 6.8).
 */
export function minimumCycleBasis(c) {
  const nu = c.cyclomaticNumber();
  if (nu <= 0) return [];
  const byName = new Map(c.reactions.map((r) => [r.name, r]));
  const candidates = [];
  for (const r of c.reactions) {
    const path = shortestPathExcluding(c, r.dst, r.src, r.name);
    if (!path) continue;
    const edges = [[r.name, +1], ...path];
    candidates.push(new Cycle(`cyc_${r.name}`, edges, nodesOf(edges, byName, r.src)));
  }
  candidates.sort((a, b) => a.length - b.length || a.name.localeCompare(b.name));

  // Keep only cycles independent over GF(2), using BigInt bitsets.
  const idx = new Map(c.reactions.map((r, i) => [r.name, i]));
  const basis = [];
  const chosen = [];
  for (const cy of candidates) {
    let vec = 0n;
    for (const [ename] of cy.edges) vec ^= 1n << BigInt(idx.get(ename));
    let red = vec;
    for (const bv of basis) {
      const x = red ^ bv;
      if (x < red) red = x;
    }
    if (red !== 0n) {
      basis.push(red);
      basis.sort((a, b) => (a < b ? 1 : a > b ? -1 : 0));
      chosen.push(cy);
    }
    if (chosen.length === nu) break;
  }
  return chosen;
}

/** Fundamental basis from a spanning forest -- the comparison basis. */
export function fundamentalCycleBasis(c) {
  const adj = adjacency(c);
  const parent = new Map();
  const seen = new Set();
  const treeEdges = new Set();
  for (const root of c.species.keys()) {
    if (seen.has(root)) continue;
    seen.add(root);
    parent.set(root, null);
    const stack = [root];
    while (stack.length) {
      const u = stack.pop();
      for (const [v, ename] of adj.get(u) || []) {
        if (!seen.has(v)) {
          seen.add(v);
          parent.set(v, [u, ename]);
          treeEdges.add(ename);
          stack.push(v);
        }
      }
    }
  }
  const byName = new Map(c.reactions.map((r) => [r.name, r]));
  const cycles = [];
  for (const r of c.reactions) {
    if (treeEdges.has(r.name)) continue;
    const path = treePath(r.dst, r.src, parent, byName);
    if (!path) continue;
    const edges = [[r.name, +1], ...path];
    cycles.push(new Cycle(`cyc_${r.name}`, edges, nodesOf(edges, byName, r.src)));
  }
  return cycles;
}

function treePath(a, b, parent, byName) {
  const ancA = [];
  let node = a;
  const seenA = new Map([[a, 0]]);
  while (parent.get(node)) {
    const [u, ename] = parent.get(node);
    ancA.push([ename, node, u]);
    node = u;
    seenA.set(node, ancA.length);
  }
  const pathB = [];
  node = b;
  while (!seenA.has(node)) {
    const p = parent.get(node);
    if (!p) return null;
    const [u, ename] = p;
    pathB.push([ename, node, u]);
    node = u;
  }
  const meet = node;
  const out = [];
  for (const [ename, child] of ancA.slice(0, seenA.get(meet))) {
    out.push([ename, byName.get(ename).src === child ? +1 : -1]);
  }
  for (let i = pathB.length - 1; i >= 0; i--) {
    const [ename, , par] = pathB[i];
    out.push([ename, byName.get(ename).src === par ? +1 : -1]);
  }
  return out;
}

// ---------------------------------------------------------------------
// Cycle sums and the two floors
// ---------------------------------------------------------------------

export function cycleSum(c, cy) {
  const byName = new Map(c.reactions.map((r) => [r.name, r]));
  let total = 0;
  for (const [ename, sign] of cy.edges) total += sign * c.deltaMu(byName.get(ename));
  return total;
}

export function potentialRange(c, cy) {
  const uniq = new Set(cy.nodes);
  let m = 0;
  for (const n of uniq) m = Math.max(m, Math.abs(c.mu.get(n) ?? 0));
  return m;
}

export function gamma(k, u = MACH_U) {
  const denom = 1 - k * u;
  if (denom <= 0) throw new KernelError("gamma: k*u >= 1, bound unusable");
  return (k * u) / denom;
}

/** eps_num = gamma_{2L} * 2 L Lambda  (Thm 3.2) */
export function numericalFloor(c, cy) {
  const L = cy.length;
  return gamma(2 * L) * 2 * L * potentialRange(c, cy);
}

/**
 * eps_data = z * sqrt(sum sigma_i^2), or null when any sigma is absent.
 * Returning null rather than defaulting is deliberate: a fabricated
 * uncertainty turns an unanswerable question into a confident answer.
 */
export function dataFloor(c, cy, z = Z_ALPHA) {
  let total = 0;
  for (const n of new Set(cy.nodes)) {
    const s = c.species.get(n)?.sigma;
    if (s === null || s === undefined) return null;
    total += s * s;
  }
  return z * Math.sqrt(total);
}

export function computeTolerance(c, cy, z = Z_ALPHA) {
  const numerical = numericalFloor(c, cy);
  const data = dataFloor(c, cy, z);
  return {
    loop: cy.name,
    numerical,
    data,
    star: data === null ? numerical : Math.max(numerical, data),
    dataAvailable: data !== null,
  };
}

export function computeHolonomy(c, cy) {
  return {
    loop: cy.name,
    value: cycleSum(c, cy),
    length: cy.length,
    potentialRange: potentialRange(c, cy),
  };
}

/** The three-valued classification of Definition 5.4. */
export function admit(h, t) {
  const a = Math.abs(h.value);
  let label;
  if (a <= t.numerical) label = CONSISTENT;
  else if (t.data !== null && a <= t.data) label = UNDECIDABLE;
  else if (t.data === null) label = UNDECIDABLE; // no sigma: cannot resolve
  else label = INCONSISTENT;
  return { label, holonomy: h, tolerance: t };
}

/**
 * Edges lying on EVERY flagged cycle. The basis-independent content of
 * a positive result: naming a single loop is not basis-independent.
 */
export function witnessSet(flagged) {
  if (!flagged.length) return new Set();
  let out = new Set(flagged[0].edgeNames());
  for (const cy of flagged.slice(1)) {
    const e = cy.edgeNames();
    out = new Set([...out].filter((x) => e.has(x)));
  }
  return out;
}

/**
 * Inject a genuine defect. It must be on an EDGE: shifting a node
 * potential leaves the edge data the gradient of the shifted potential,
 * so every cycle sum stays exactly zero. Only an edge offset makes the
 * data fail to be a gradient, which is what inconsistency means.
 */
export function perturbEdge(c, reactionName, delta) {
  if (!c.reactions.some((r) => r.name === reactionName)) {
    throw new KernelError(`perturb_edge: unknown reaction '${reactionName}'`);
  }
  const out = c.copy();
  out.edgeOffset.set(reactionName, (out.edgeOffset.get(reactionName) || 0) + delta);
  return out;
}

/** Node perturbation -- provided, but it cannot create inconsistency. */
export function perturbNode(c, speciesName, delta) {
  if (!c.mu.has(speciesName)) {
    throw new KernelError(`perturb: unknown species '${speciesName}'`);
  }
  const out = c.copy();
  out.mu.set(speciesName, out.mu.get(speciesName) + delta);
  return out;
}
